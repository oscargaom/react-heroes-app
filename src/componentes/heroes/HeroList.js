import React, { useMemo } from 'react'
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher'
import { HeroCard } from './HeroCard';

export const HeroList = ({publisher}) => {

    
    /*  Se recomienda que en este tipo de llamados se haga uso del hook useMemo
    para guardar los valores que retorna la función o bien un api, para que 
    cada vez que se cargue la página no se haga un llamado extra y esto afecte 
    el performance de la aplicación. el hook useMemo guardará el último valor 
    del llamado y solamente se volverá a ejecutar si alguna de las variables 
    de las cuales tiene depdencia sufre algún cambio o modificación.
    */

    // const heroes = getHeroesByPublisher(publisher);
    const heroes =  useMemo(() => getHeroesByPublisher(publisher), [publisher]);

    // console.log(heroes);

    return (
        <div className="card-columns animate__animated animate__pulse">
            {
                heroes.map( hero => 
                    <HeroCard key={hero.id} {...hero}/>
                )
            }
        </div>
    )
}
