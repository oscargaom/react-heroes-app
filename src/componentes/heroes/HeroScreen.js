import React, { useMemo } from 'react'
import { useParams, Link, Redirect } from 'react-router-dom'
import { getHeroById } from '../../selectors/getHeroById';

export const HeroScreen = ({history}) => {

    /*  useParams es un hook que nos proporciona react-router-dom y nos
        ayuda a obtener los parámetros de las urls (query params). 
        Como en este caso hacemos uso del parámetro heroId en la 
        url: /hero/:heroId, este hook nos ayudará a obtenerlo de manera sencilla.
    */
    const params = useParams();

    const { heroId } = params;

    // console.log(heroId);

    // const hero = getHeroById(heroId);
    
    /*  Se recomienda que en este tipo de llamados se haga uso del hook useMemo
        para guardar los valores que retorna la función o bien un api, para que 
        cada vez que se cargue la página no se haga un llamado extra y esto afecte 
        el performance de la aplicación. el hook useMemo guardará el último valor 
        del llamado y solamente se volverá a ejecutar si alguna de las variables 
        de las cuales tiene depdencia sufre algún cambio o modificación.
    */
    const hero = useMemo(() => getHeroById(heroId), [heroId]);

    // console.log(hero);

    const handleReturn = () => {
        
        // console.log(history);
        
        /*  history.length contine el histórico del número de páginas sobre 
        las que hemos navegado, de modo que si son pocas significa que 
        copiamos la url y comenzamos a navegar desde ahí y no desde un login.
        */
       if (history.length <= 2) {
           history.push("/");
        } else {
            history.goBack();
        }
    };

    if (!hero) {
        // console.log('<Redirect to="/"');
        return <Redirect to="/" />
    }

    const { superhero, publisher, alter_ego,
        first_appearance, characters } = hero;

    // console.log(superhero);
    // console.log(publisher);
    // console.log(alter_ego);
    // console.log(first_appearance);
    // console.log(characters);    

    return (
        <div className="row mt-5">
            <div className="col-4">
                <img src={`../assets/heroes/${heroId}.jpg`} alt={superhero} 
                    className="img-thumbnail card-img animate__animated animate__zoomInDown" />
            </div>
            <div className="col-8 animate__animated animate__fadeIn">
                <h3>{superhero}</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item">
                        <b>Alter ego: </b> {alter_ego}
                    </li>
                    <li className="list-group-item">
                        <b>Publisher: </b> {publisher}
                    </li>
                    <li className="list-group-item">
                        <b>First Appearance: </b> {first_appearance}
                    </li>
                </ul>

                <h5>Characters</h5>
                <p>{characters}</p>

                <button
                    className="btn btn-outline-info"
                    onClick={handleReturn}
                >
                    Return
                </button>
            </div>
        </div>
    )
}
