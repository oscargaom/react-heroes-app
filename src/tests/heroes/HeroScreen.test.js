import React from 'react';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';

import { HeroScreen } from '../../componentes/heroes/HeroScreen';
import { historyMockLgthEq2, historyMockLgthEq10 } from '../fixtures/authContextValue';


describe('Pruebas en el componente <HeroScreen />', () => {


    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('debe mostrarse el componente <Link to="/" /> si no hay argumentos en la url ', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <HeroScreen history={historyMockLgthEq2} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('Redirect').exists()).toBe(true);
    });

    test('debe de mostrar un hero si el parámetro existe en el query string y se encuentra en la ruta', () => {

        /*  initialEntries={['/hero/marvel-spider']} reemplaza el valor marvel-spider en el :heroId del path de
            <Route path="/hero/:heroId" .... /> , pero es importante que en ambos inicien con /hero/
            Route le proporciona todas las props a <HeroScreen /> a través de la función flecha en component={...}
        */
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                {/* <HeroScreen history={historyMockLgthEq10} /> */}
                <Route
                    path="/hero/:heroId"
                    component={() => <HeroScreen history={historyMockLgthEq2} />} />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();
        // console.log(wrapper.html());
        expect(wrapper.find('.row').exists()).toBe(true);
        expect(historyMockLgthEq2.goBack).not.toHaveBeenCalled();
        expect(historyMockLgthEq2.push).toBeCalledTimes(1);
        expect(historyMockLgthEq2.push).toHaveBeenCalledWith('/');
    });

    test('debe de regresar a la página anteior mediante history.goBack()', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route 
                    path='/hero/:heroId' 
                    component={ () => <HeroScreen history={historyMockLgthEq10} />} >
                    {/* <HeroScreen history={historyMockLgthEq10} /> */}
                </Route>
            </MemoryRouter>
        );

        wrapper.find('button').props().onClick();
        expect(historyMockLgthEq10.push).not.toHaveBeenCalled();
        expect(historyMockLgthEq10.goBack).toHaveBeenCalledTimes(1);
    });

    test('debe llamar <Redirect to="/" /> si el hero no existe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider7']}>
                <Route
                    path='/hero/:heroId'
                    component={() => <HeroScreen history={historyMockLgthEq2} />}>
                </Route>
            </MemoryRouter>
        );

        /*  El <Redirect to="/" /> regresa una cadena vacía por eso no podemos buscar 
            por el nombre del componente, pero si podemos hacer la validación de la
            cadena vacía. */
        // expect(wrapper.find('Redirect').exists()).toBe(true);
        expect(wrapper.text()).toBe('');
    });

})


