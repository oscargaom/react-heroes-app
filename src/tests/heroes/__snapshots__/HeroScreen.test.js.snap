// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <HeroScreen /> debe mostrarse el componente <Link to="/" /> si no hay argumentos en la url  1`] = `""`;
